### ISSUE TITLE
Support configurable serialization inclusion for Jackson

### ISSUE THREAD

**Issue by rpatrick00** [joelittlejohn/jsonschema2pojo#629]
We have been using Jackson 2.7.6 with the objects created by jsonschema2pojo with the following mapper configuration:

`public static ObjectMapper getJacksonObjectMapper() {
    ObjectMapper mapper = new ObjectMapper();
    mapper.configure(SerializationFeature.INDENT_OUTPUT, true);
    mapper.configure(SerializationFeature.FAIL_ON_EMPTY_BEANS, false);
    mapper.configure(SerializationFeature.WRITE_EMPTY_JSON_ARRAYS, false);
    mapper.setSerializationInclusion(JsonInclude.Include.NON_EMPTY);
    return mapper;
}`

This is working perfectly and eliminates empty JSON arrays from being included in the serialized form.  

We have recently moved from Jackson 2.7.6 to Jackson 2.8.3, which has deprecated the SerializationFeature.WRITE_EMPTY_JSON_ARRAYS feature.  I raised issue [`102`](https://github.com/FasterXML/jackson-annotations/issues/102) on Jackson annotations and what they are telling me is that the class-level @JsonInclude(JsonInclude.Include.NON_NULL) your tool includes is overriding the mapper-level configuration of mapper.setSerializationInclusion(JsonInclude.Include.NON_EMPTY).  A simple test case is attached to the Jackson issue.

As such, I really need a solution to eliminate the need for the deprecated Jackson feature with your tool that allows me to prevent the serialization of empty JSON arrays.

**Comment by joelittlejohn** [comment]
Hi. If you don't want the types to create empty collections by default you can turn these off, see https://joelittlejohn.github.io/jsonschema2pojo/site/0.4.27/generate-mojo.html#initializeCollections.

If you set these values yourself then you're going to see them serialized.

**Comment by rpatrick00** [comment]
Well, there are side-effects of doing what you suggest, right?  Now, every time I access a collection I need to test for null.  I don't understand why I need to live with the implications to change the serialization behavior...

**Comment by thachhoang** [comment]
If you don't want to use `initializeCollections`, you can either:
- Edit the schema to make those array elements not "required"
- Use a [custom annotator](https://joelittlejohn.github.io/jsonschema2pojo/site/0.4.27/generate-mojo.html#customAnnotator) that doesn't add `JsonInclude.Include.NON_NULL` to your classes:

  ```java
  public class CustomJackson2Annotator extends Jackson2Annotator {
    @Override
    public void propertyInclusion(JDefinedClass clazz, JsonNode schema) {
      // or leave this method blank for no annotations
      clazz.annotate(JsonInclude.class).param("value", JsonInclude.Include.NON_EMPTY);
    }
  }
  ```

**Comment by rpatrick00** [comment]
Thanks.  The elements are already not required but were still showing up.  The custom annotator seems like a hammer but should certainly work.  I was hoping for a more configuration-based approach.  I glanced through the code this afternoon looking at how to add a configuration option to control the class-level annotations but didn't see an easy way of getting the configuration accessible from code that creates the annotation.

I did find an alternative Mapper-level mechanism that allows me to change the Jackson default behavior by type that allowed me to work around the problem again without using the deprecated feature.

**Comment by joelittlejohn** [comment]
Hi @rpatrick00 yes, I see.

I think we can solve this with two changes:

1. Add a configuration option like `jacksonSerializationInclusion` that allows the inclusion mode to be set

2. Add a constructor to the annotators that accepts the GenerationConfig. When creating an annotator instance, we can use this constructor if it is present and use the default constructor if it is not. This should make it a backward compatible change.

**Comment by joelittlejohn** [comment]
It sounds like you have a couple of workarounds here though. A custom annotator (as @thachhoang suggests) or your alternative Jackson configuration that you have found. Could you share that configuration option? It would be useful for others that find this ticket. Cheers.

**Comment by rpatrick00** [comment]
I added default settings by type in my mapper configuration:

        mapper.configOverride(List.class).setInclude(
            JsonInclude.Value.construct(JsonInclude.Include.NON_EMPTY, null));
        mapper.configOverride(Set.class).setInclude(
            JsonInclude.Value.construct(JsonInclude.Include.NON_EMPTY, null));

**Comment by ramdog** [comment]
Thanks for posting that solution @rpatrick00. For Jackson 2.7, you can achieve the same thing with:

```java
mapper.configure(SerializationFeature.WRITE_EMPTY_JSON_ARRAYS, false);
```

**Comment by thachhoang** [comment]
Fixed by #671.
